src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -0,0 +1,174 @@
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * Worker class to collect insert size metrics.
+ */
+public final class InsertSizeMetricsCollectorSpark implements Serializable{
+    private static final long serialVersionUID = 1L;
+
+    private final JavaRDD<Integer> lengthRDD;
+    private final InsertSizeMetrics metrics = new InsertSizeMetrics();
+    private Histogram<Integer> htslibHistogram;
+    private final long pairsCount;
+
+    /**
+     * Constructor and do collection work.
+     * @param filteredReads          reads that pass filters
+     * @param samRgRecords           a list of read groups
+     * @param HistogramMADTolerance  MAD tolerance when producing histogram plot
+     */
+    // TODO: * @param accumLevels            accumulation level {SAMPLE, LIBRARY, READGROUP}
+    public InsertSizeMetricsCollectorSpark(final JavaRDD<GATKRead> filteredReads,
+                                           //final Set<MetricAccumulationLevel> accumLevels,
+                                           final List<SAMReadGroupRecord> samRgRecords,
+                                           final double HistogramMADTolerance) {
+
+        lengthRDD = filteredReads.map(read -> Math.abs(read.getFragmentLength()) );
+        lengthRDD.cache();
+
+        metrics.READ_PAIRS = filteredReads.count();
+        this.pairsCount = metrics.READ_PAIRS;
+
+        collectLevelInfo();
+
+        collectMedianAndMAD();
+
+        collectSymmetricalBinWidth();
+
+        trimHistogramAndCollectMeanAndSD(HistogramMADTolerance);
+    }
+
+    public final InsertSizeMetrics getMetrics(){
+        return metrics;
+    }
+
+    public final Histogram<Integer> getHitogram(){
+        return htslibHistogram;
+    }
+
+    // simply return an double array from an RDD of integers for using Apache DescriptiveStatistics
+    private double[] getRDDArray(final JavaRDD<Integer> rdd){
+        final List<Integer> tempList = rdd.collect();
+        final double lengthArr[] = new double[tempList.size()];
+        for(int i=0; i<lengthArr.length; ++i){
+            lengthArr[i] = tempList.get(i);
+        }
+        return lengthArr;
+    }
+
+    private void collectMedianAndMAD(){
+
+        double lengthArr[] = getRDDArray(lengthRDD);
+
+        DescriptiveStatistics percent = new DescriptiveStatistics(lengthArr);
+
+        metrics.MIN_INSERT_SIZE     = (int) percent.getMin();
+        metrics.MAX_INSERT_SIZE     = (int) percent.getMax();
+
+        metrics.MEDIAN_INSERT_SIZE  = percent.getPercentile(50.0);
+
+        final int median = (int) metrics.MEDIAN_INSERT_SIZE;
+        lengthArr = getRDDArray( lengthRDD.map(length -> Math.abs(length - median)) );  // reuse
+        percent = new DescriptiveStatistics(lengthArr);                                 // reuse
+        metrics.MEDIAN_ABSOLUTE_DEVIATION = percent.getPercentile(50.0);
+    }
+
+    private void collectSymmetricalBinWidth(){
```

Comment:
The part of this method that takes a `SortedMap<Integer, Long> hist ` and returns a `long[]` should be extracted to a separate method and unit-tested separately